client.go

```diff
@@ -636,23 +630,30 @@ func (c *Client) Do(req *Request) (*http.Response, error) {
 		}
 
 		wait := c.Backoff(c.RetryWaitMin, c.RetryWaitMax, i, resp)
-		desc := fmt.Sprintf("%s %s", req.Method, req.URL)
-		if code > 0 {
-			desc = fmt.Sprintf("%s (status: %d)", desc, code)
-		}
 		if logger != nil {
+			var code int // HTTP response code
+			if resp != nil {
+				code = resp.StatusCode
+			}
+			var desc string
+			if code > 0 {
+				desc = fmt.Sprintf("%s (status: %d)", desc, code)
+			} else {
+				desc = fmt.Sprintf("%s %s", req.Method, req.URL)
+			}
 			switch v := logger.(type) {
 			case LeveledLogger:
 				v.Debug("retrying request", "request", desc, "timeout", wait, "remaining", remain)
 			case Logger:
 				v.Printf("[DEBUG] %s: retrying in %s (%d left)", desc, wait, remain)
 			}
 		}
+		timer := time.NewTimer(wait)
```

Comment:
This adds (a small amount of) noise to the intent here. The only case we return early is when the context is canceled. It's true that the channel will be eligible for GC sooner, so I suppose if you had set an extremely long `RetryWaitMax` or something, this could give a very small benefit. Overall I probably wouldn't have changed this myself as I think it reads better as a one-liner.

Reply:
Yeah, I agree it's a bit more noisy but I think it's good for such a widely used library to be as efficient as possible. Only very few people will need to look at it, but a lot of the users will benefit from improvements.

Reply:
Cool, I agree with that. 👍